Migrate to server routing

hudson-newey opened the pull request:
# Migrate to server routing

This enables [hybrid rendering](https://angular.dev/guide/hybrid-rendering) and allows us to specify what rendering strategy we should use on a per-page basis through the `PageInfo`.

---

We might want to expose this through the `StrongRoute` instead of the `PageInfo` because `ServerRoute`'s support a lot of other cool features such as specific headers for certain routes, status codes, etc....

https://angular.dev/guide/hybrid-rendering#setting-headers-and-status-codes

Additionally, putting all of this information on the `PageInfo` leaks server routing information to the client which we might not want.

## Changes

- Migrates to the app server router.
- Migrate the home page to use pre-rendering (ssg) and defer the loading of "example project" cards to the client.
	- this was done more or less to test that the server routing / `renderMode` was working correctly. Let me know if you want this reverted, but I think it's a nice improvement since it improves the load times, and the "example project" cards are typically off screen on first load, so by the time the user scrolls to them, they would have loaded in by csr.
- Replaces `REQUEST` and `RESPONSE` injection tokens with Angular native SSR tokens.
- Do not provide `APP_CONFIG` provider twice (we were providing it through the express provides and the `config.service`.
- Fixes console errors being thrown when page title route models could not be accessed due to un-authenticated sessions.
- Do not show "bad" website status warnings if the website status has not been fetched.

## Issues

Fixes: #2290

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

hudson-newey committed:
Add renderMode to PageInfo

hudson-newey committed:
Add server routing

atruskie commented on src/app/components/projects/pages/new/new.component.ts:
```diff
@@ -62,6 +63,7 @@ ProjectNewComponent.linkToRoute({
   category: projectsCategory,
   pageRoute: newProjectMenuItem,
   menus: { actions: List(projectsMenuItemActions) },
+  renderMode: RenderMode.Client,
```
Add a comment explaining why this route in particular is client rendering?

hudson-newey replied:
Reverted

atruskie commented on src/app/helpers/page/pageInfo.ts:
```diff
@@ -28,6 +27,12 @@ export interface IPageInfo extends Data {
+
+  /**
+   * Describes where the page will be rendered.
+   * @default RenderMode.Server
```
add a comment about how our linkToRoute(?) function plumbs this into the route data for the page

atruskie commented on src/server.new.ts:
i assume this is WIP and we'll just keep one server.ts?

hudson-newey committed:
Correct seperation of client and server routes

hudson-newey committed:
Re-add static routes

hudson-newey committed:
Fix page titles errors when not authenticated

This was causing ssr to throw a lot of errors because ssr would always
be unauthenticated and unable to access private projects.

hudson-newey committed:
Do not declare server routes in client

hudson-newey committed:
Re-instate ssg providers

hudson-newey committed:
Clean up unused express tokens

We previously had three providers that were provided through the
CommonModule's express bootstrap, however, this has now been removed.

Some of the injection tokens have been replaced with angular native
tokens e.g. REQUEST, and RESPONSE_INIT. However, some like the
APP_CONFIG are custom to our solution.

Having looked through the code, I have determined that we were actually
reading the config twice in SSR (once through the express bootstrap and
once through the config.service).
This has now been unified to just use the config.service.

hudson-newey committed:
Ship X-Frame-Options with all requests

Not just client pages

hudson-newey committed:
Add pageRouting tests

hudson-newey committed:
Fix CardComponent mock causing no defer support

hudson-newey committed:
Revert defer block

ng-mocks doesn't support defer blocks, which makes it really hard to
test components with defer blocks.

hudson-newey committed:
Revert defer block testing in home.component.spec

atruskie commented on src/app/test/fakes/PageInfo.ts:
```diff
@@ -1,14 +1,20 @@
+      RenderMode.Server,
+      RenderMode.Client,
+      RenderMode.Prerender,
+    ]),
```
probably don't want to randomize this, it could create flaky tests that render different ways? It depends how often we use this pageinfo generator/helper...

hudson-newey replied:
We use the `PageInfo` generator/helper a lot.

Most of our Testbed's don't actually have a router, and those that do provide their own router, meaning that the mocked routes don't use `StrongRoute`'s or `PageInfo` at all.

e.g.
https://github.com/QutEcoacoustics/workbench-client/blob/bfee1a82c92a8f56d2b192dbb42cdb55191c1545/src/app/directives/strongRoute/strong-route-active.directive.spec.ts#L38-L48

Additionally, all of our tests are client-side only.\
We don't have any ssr or e2e tests.
Meaning that it's not possible to provide `ServerRoute`'s (which is the type of route that takes a `RenderMode`. Client side Route's cannot take a `RenderMode`).

hudson-newey replied:
![image](https://github.com/user-attachments/assets/f797e8b7-aebb-4b1d-b963-a71f44ed16da)

hudson-newey committed:
Use consistent RenderMode for PageInfo mocks
